File: backend-v2/app/api/routers/admin/configuration.py
from fastapi import APIRouter, Depends

from app.core.dependencies import repository, service
from app.core.security import actor_id, require_admin_token
from app.repositories.sqlite import SQLiteRepository
from app.schemas import (
    EventLocation,
    FeatureLink,
    FeatureToggleInput,
    NavigationLink,
    OSMSearchResult,
    SiteConfig,
)
from app.services.hackathon import HackathonService

router = APIRouter(dependencies=[Depends(require_admin_token)])


@router.get("/navigation-links", response_model=list[NavigationLink], response_model_by_alias=True)
def admin_navigation_links(repo: SQLiteRepository = Depends(repository)) -> list[NavigationLink]:
    return repo.list_navigation_links(include_disabled=True)


@router.post(
    "/navigation-links",
    status_code=201,
    response_model=NavigationLink,
    response_model_by_alias=True,
)
def create_navigation_link(
    input: NavigationLink,
    actor: str = Depends(actor_id),
    svc: HackathonService = Depends(service),
) -> NavigationLink:
    return svc.create_navigation_link(actor, input)


@router.get("/feature-links", response_model=list[FeatureLink], response_model_by_alias=True)
def admin_feature_links(repo: SQLiteRepository = Depends(repository)) -> list[FeatureLink]:
    return repo.list_feature_links(include_disabled=True)


@router.post(
    "/feature-links",
    status_code=201,
    response_model=FeatureLink,
    response_model_by_alias=True,
)
def create_feature_link(
    input: FeatureLink,
    actor: str = Depends(actor_id),
    svc: HackathonService = Depends(service),
) -> FeatureLink:
    return svc.create_feature_link(actor, input)


@router.patch(
    "/feature-links/{feature_id}",
    response_model=FeatureLink,
    response_model_by_alias=True,
)
def update_feature_link(
    feature_id: str,
    input: FeatureToggleInput,
    actor: str = Depends(actor_id),
    svc: HackathonService = Depends(service),
) -> FeatureLink:
    return svc.set_feature_enabled(actor, feature_id, input.enabled)


@router.get("/locations/search", response_model=list[OSMSearchResult], response_model_by_alias=True)
def search_locations(q: str, svc: HackathonService = Depends(service)) -> list[OSMSearchResult]:
    return svc.search_locations(q)


@router.get("/event-location", response_model=EventLocation, response_model_by_alias=True)
def admin_event_location(repo: SQLiteRepository = Depends(repository)) -> EventLocation:
    return repo.get_event_location()


@router.put("/event-location", response_model=EventLocation, response_model_by_alias=True)
def update_event_location(
    input: EventLocation,
    actor: str = Depends(actor_id),
    svc: HackathonService = Depends(service),
) -> EventLocation:
    return svc.update_event_location(actor, input)


@router.put("/site-config", response_model=SiteConfig, response_model_by_alias=True)
def update_site_config(
    input: SiteConfig,
    actor: str = Depends(actor_id),
    svc: HackathonService = Depends(service),
) -> SiteConfig:
    return svc.update_site_config(actor, input)

File: frontend/web/lib/api/configuration.ts
import { request } from "./client";
import type {
  AccommodationRequest,
  EventLocation,
  FeatureLink,
  NavigationLink,
  OSMSearchResult,
  SecretKeyList,
  SiteConfig,
} from "./types";

export const configurationApi = {
  navigationLinks: () => request<NavigationLink[]>("/api/navigation-links"),
  adminNavigationLinks: () => request<NavigationLink[]>("/api/admin/navigation-links", { admin: true }),
  createNavigationLink: (input: Pick<NavigationLink, "title" | "description" | "url"> & { sortOrder?: number }) =>
    request<NavigationLink>("/api/admin/navigation-links", {
      admin: true,
      method: "POST",
      body: JSON.stringify(input),
    }),
  featureLinks: () => request<FeatureLink[]>("/api/feature-links"),
  adminFeatureLinks: () => request<FeatureLink[]>("/api/admin/feature-links", { admin: true }),
  updateFeatureEnabled: (id: string, enabled: boolean) =>
    request<FeatureLink>(`/api/admin/feature-links/${id}`, {
      admin: true,
      method: "PATCH",
      body: JSON.stringify({ enabled }),
    }),
  siteConfig: () => request<SiteConfig>("/api/site-config"),
  updateSiteConfig: (input: Partial<SiteConfig>) =>
    request<SiteConfig>("/api/admin/site-config", {
      admin: true,
      method: "PUT",
      body: JSON.stringify(input),
    }),
  eventLocation: () => request<EventLocation>("/api/event-location"),
  adminEventLocation: () => request<EventLocation>("/api/admin/event-location", { admin: true }),
  searchLocations: (query: string) =>
    request<OSMSearchResult[]>(`/api/admin/locations/search?q=${encodeURIComponent(query)}`, { admin: true }),
  updateEventLocation: (input: EventLocation) =>
    request<EventLocation>("/api/admin/event-location", {
      admin: true,
      method: "PUT",
      body: JSON.stringify(input),
    }),
  accommodation: () => request<AccommodationRequest>("/api/accommodation"),
  adminAccommodationRequests: () =>
    request<AccommodationRequest[]>("/api/admin/accommodation-requests", { admin: true }),
  updateAccommodation: (input: Pick<AccommodationRequest, "selections" | "otherDetail">) =>
    request<AccommodationRequest>("/api/accommodation", {
      method: "PUT",
      body: JSON.stringify(input),
    }),
  emailSecrets: () => request<SecretKeyList>("/api/admin/email-secrets", { admin: true }),
  setEmailSecret: (key: string, value: string) =>
    request<{ status: string }>(`/api/admin/email-secrets/${key}`, {
      admin: true,
      method: "PUT",
      body: JSON.stringify({ value }),
    }),
  deleteEmailSecret: (key: string) =>
    request<{ status: string }>(`/api/admin/email-secrets/${key}`, {
      admin: true,
      method: "DELETE",
    }),
};
